### PR TITLE
Donk Pocket Cooling 2: The Reworkening

### PR DESCRIPTION
**Pull request by Coolrune206**
## What Does This PR Do

Revises https://github.com/ParadiseSS13/Paradise/pull/20417 on feedback both from that PR and from Coding Chat, particularly Denthamos. Warm donk pockets now cool over time, with a timer of seven minutes until they are cold. However, they may be re-heated in a microwave. Additionally, there is now a var for food recipes that allows for no duplication of output if the machine is upgraded, which is used by normal donk pockets being warmed up and warm/cold donk pockets being reheated.
The Donk Pocket Box has been slightly reworked, it now can only hold donk pockets (6 of them) but while warm donk pockets are inside the box their cooling timer is frozen - if you put one that will cool in 5 minutes in, it will stop its timer until it is removed at which point the 5 minute timer continues.
Also, a sprite for the Syndie Donk Pocket box was added while I was here, because I was originally going to rework the sprite of the original Donk Pocket box to use the /fancy/ subtype instead of the /box/ subtype so I could have it visually display its contents, but then I realized that would require a mapping change on all maps and a bunch of space ruins, so I've left that out of this PR - I may approach it in the future.
Donk Pocket Boxes can be crafted using a regular box, a piece of metal, and a piece of plastic - for the insulation.
Cold Donk pockets give 5 diluted omnizine instead of 15 omnizine that donk pockets have.
This change does not affect syndie donk pockets outside of the change to the box's sprite.

## Why It's Good For The Game

As stated in the original PR, Donk Pockets are incredibly strong healing options, healing most damage types (Brute, Burn, Toxin, Oxygen) and reducing losebreath, and the amount of omnizine in donk pockets is enough to heal 150 damage per donk. This is a lot considering you can fit seven in a box and they'd never cool, despite the game stating they would. However, just having them cool every seven minutes would be tedium as you go to re-heat them, so the box freezing the timer on the donks seems like a fair trade-off - you sacrifice a space in your inventory for the donk pocket box (which only holds the donk pockets) in exchange for its very good healing properties.

## Images of changes

Simple sprite change, the syndicate donk pocket box:
image

## Testing

Ensured Donk Pockets cool after the timer ends, ensured they can be re-heated, ensured putting them in a box freezes the timer, ensured both donk heating recipes did not duplicate the donks, ensured regular recipes (like making donk pockets from meatballs and dough) duplicated properly in an upgraded microwave. Should be good.

## Changelog
:cl:
tweak: Donk Pockets cool over seven minutes, becoming cold and less effective
tweak: Donk Pocket Boxes insulate the Donk Pockets, preventing them from cooling while inside
tweak: Heating Donk Pockets no longer create additional Warm Donk Pockets when microwave is upgraded
imageadd: Syndie Donk Pocket Box sprite
/:cl:

### CODE REVIEW

File: code/datums/recipe.dm [hunk 1]
```diff
@@ -38,6 +38,7 @@
 	var/list/items // example: =list(/obj/item/crowbar, /obj/item/welder) // place /foo/bar before /foo
 	var/result //example: = /obj/item/reagent_containers/food/snacks/donut
 	var/time = 100 // 1/10 part of second
+	var/duplicate = TRUE // Whether or not an upgraded kitchen machine will create more products using the same amount of ingredients
```
Comment (lewcc) [inline code/datums/recipe.dm:41]: Could you autodoc this comment? I know it wouldn't be consistent with the rest of the comments but it's still better practice.

File: code/modules/food_and_drinks/food/foods/misc_food.dm [hunk 2]
```diff
@@ -114,9 +114,41 @@
...
+	warm = FALSE
+
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/on_enter_storage(obj/item/storage/S as obj)
+	. = ..()
+	if(istype(S, /obj/item/storage/box/donkpockets) && cool_timer)
+		deltimer(cool_timer)
+		subtracted_time += (world.time - heated_at)
+		cool_timer = null
+
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/on_exit_storage(obj/item/storage/S as obj)
```
Comment (lewcc) [inline code/modules/food_and_drinks/food/foods/misc_food.dm:142]:
```suggestion
/obj/item/reagent_containers/food/snacks/warmdonkpocket/on_exit_storage(obj/item/storage/S)
```

File: code/modules/food_and_drinks/food/foods/misc_food.dm [hunk 3]
```diff
@@ -114,9 +114,41 @@
...
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/proc/donk_heat()
+	heated_at = world.time
+	cool_timer = addtimer(CALLBACK(src, PROC_REF(donk_cool)), 420 SECONDS - subtracted_time, TIMER_UNIQUE|TIMER_STOPPABLE)
+
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/proc/donk_cool()
+	name = "cold Donk-pocket"
+	desc = "The food of choice for the seasoned traitor. This one is cold."
+	warm = FALSE
+
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/on_enter_storage(obj/item/storage/S as obj)
```
Comment (lewcc) [inline code/modules/food_and_drinks/food/foods/misc_food.dm:135]: We don't do this anymore
```suggestion
/obj/item/reagent_containers/food/snacks/warmdonkpocket/on_enter_storage(obj/item/storage/S)
```

File: code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm [hunk 4]
```diff
@@ -288,11 +288,15 @@
...
+		var/portions = recipe.duplicate ? efficiency : 1
+		var/reagents_per_serving = temp_reagents.total_volume / portions
+		if(recipe.duplicate)
+			for(var/i in 1 to portions) // Extra servings when upgraded, ingredient reagents split equally
+				var/obj/cooked = new recipe.result(loc)
+				temp_reagents.trans_to(cooked, reagents_per_serving, no_react = TRUE) // Don't react with the abstract holder please
+		else
 			var/obj/cooked = new recipe.result(loc)
-			temp_reagents.trans_to(cooked, reagents_per_serving, no_react = TRUE) // Don't react with the abstract holder please
+			temp_reagents.trans_to(cooked, temp_reagents.total_volume, no_react = TRUE) // Don't react with the abstract holder please
```
Comment (lewcc) [inline code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm:299]: I don't think you need to branch on `if(recipe.duplicate)` here. If we aren't duplicating, `reagents_per_serving` should be `total_volume / 1` = `total_volume` anyway, and `for(var/i in 1 to 1)` should still run the loop body once.

**Review by lewcc — CHANGES_REQUESTED** [review]

File: code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm (hunk 4 above)
Comment (Coolrune206) [reply code/modules/food_and_drinks/kitchen_machinery/kitchen_machine.dm:299]: correct, I forgot to change that when Snowball helped me cut down on the code here. Thanks

File: code/game/objects/items/weapons/storage/boxes.dm [hunk 5]
```diff
@@ -358,17 +358,26 @@
 
 /obj/item/storage/box/donkpockets
 	name = "box of donk-pockets"
-	desc = "<B>Instructions:</B> <I>Heat in microwave. Product will cool if not eaten within seven minutes.</I>"
+	desc = "A heavy, insulated box that reads, <B>Instructions:</B> <I>Heat in microwave. Product will cool if not eaten within seven minutes. Store product in box to keep warm.</I>"
```
Comment (Sirryan2002) [inline code/game/objects/items/weapons/storage/boxes.dm:361]:
```suggestion
	desc = "A heavy, insulated box that reads, <b>Instructions:</b> <i>Heat in microwave. Product will cool if not eaten within seven minutes. Store product in box to keep warm.</i>"
```
HTML element should be lowercase

File: code/modules/food_and_drinks/food/foods/misc_food.dm [hunk 6]
```diff
@@ -114,9 +114,41 @@
...
+	var/heated_at
+	var/subtracted_time = 0
+
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/Initialize(mapload)
+	. = ..()
+	donk_heat()
+
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/proc/donk_heat()
+	heated_at = world.time
+	cool_timer = addtimer(CALLBACK(src, PROC_REF(donk_cool)), 420 SECONDS - subtracted_time, TIMER_UNIQUE|TIMER_STOPPABLE)
```
Comment (Sirryan2002) [inline code/modules/food_and_drinks/food/foods/misc_food.dm:128]:
```suggestion
	cool_timer = addtimer(CALLBACK(src, PROC_REF(donk_cool)), 7 MINUTES - subtracted_time, TIMER_UNIQUE|TIMER_STOPPABLE)
```

**Comment by Coolrune206** [comment]
the answer is no. even without any modifications to icons/obj/storage.dmi (by sacrificing the syndie donk kit changes for now), it's still decided it's FUCKING CONFLICTED.

If anyone knows how to fix this absolutely stupid problem, please do reach out to me.

**Comment by Coolrune206** [comment]
GDN is based and helped me fix the .dmi issue, les go.

File: code/modules/food_and_drinks/food/foods/misc_food.dm [hunk 7]
```diff
@@ -114,9 +114,41 @@
 	filling_color = "#DEDEAB"
 	list_reagents = list("nutriment" = 4)
 	tastes = list("meat" = 2, "dough" = 2, "laziness" = 1)
+	var/warm = TRUE
+	var/cool_timer
+	var/heated_at
+	var/subtracted_time = 0
```
Comment (S34NW) [inline code/modules/food_and_drinks/food/foods/misc_food.dm:120]: Please document these

File: code/modules/food_and_drinks/food/foods/misc_food.dm [hunk 8]
```diff
@@ -114,9 +114,41 @@
...
+	name = "cold Donk-pocket"
+	desc = "The food of choice for the seasoned traitor. This one is cold."
+	warm = FALSE
+
+/obj/item/reagent_containers/food/snacks/warmdonkpocket/on_enter_storage(obj/item/storage/S)
+	. = ..()
+	if(istype(S, /obj/item/storage/box/donkpockets) && cool_timer)
+		deltimer(cool_timer)
+		subtracted_time += (world.time - heated_at)
+		cool_timer = null
```
Comment (S34NW) [inline code/modules/food_and_drinks/food/foods/misc_food.dm:140]: Guard clause please

File: code/modules/food_and_drinks/food/foods/misc_food.dm [hunk 9]
```diff
@@ -114,9 +114,41 @@
...
+	. = ..()
+	if(warm)
+		donk_heat()
 
 /obj/item/reagent_containers/food/snacks/warmdonkpocket/Post_Consume(mob/living/M)
-	M.reagents.add_reagent("omnizine", 15)
+	if(warm)
+		M.reagents.add_reagent("omnizine", 15)
+	else
+		M.reagents.add_reagent("weak_omnizine", 5)
```
Comment (S34NW) [inline code/modules/food_and_drinks/food/foods/misc_food.dm:151]: no need for if else, just early return

**Review by S34NW — CHANGES_REQUESTED** [review]

File: code/modules/food_and_drinks/food/foods/misc_food.dm (hunk 8 above)
Comment (Coolrune206) [reply code/modules/food_and_drinks/food/foods/misc_food.dm:140]: How would I apply guard clause here? I thought that was for chains of if-else if which use early returns

Comment (Vi3trice) [reply code/modules/food_and_drinks/food/foods/misc_food.dm:144]: It should early return if it doesn't have a timer instead of always doing a type check.

**Review by S34NW — APPROVED** [review]

**Review by Sirryan2002 — APPROVED** [review]

**Review by lewcc — APPROVED** [review]

**Review by SteelSlayer — APPROVED** [review]

**Comment by DoktorBarron** [comment]
welp.
there goes Macca's exponential legally distinct pocket food product growth potential.
I'll have to somehow get to 1000 donks in a shift some other way.

**Pull request merged**